### PR TITLE
Flushing Memtree

### PR DESCRIPTION
**Pull request by Sriep**
Allows using two DBS for storing data.

### CODE REVIEW

File: mutable_tree.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ type MutableTree struct {
 	*ImmutableTree                  // The current, working tree.
 	lastSaved      *ImmutableTree   // The most recently saved tree.
 	orphans        map[string]int64 // Nodes removed by changes to working tree.
+	memversions    map[int64]bool   // The previous, saved versions of the tree in mem.
```
Comment (enlight) [inline mutable_tree.go]: `memversions` -> `memVersions`

File: mutable_tree.go [hunk 2]
```diff
@@ -364,11 +366,44 @@ func (tree *MutableTree) GetVersioned(key []byte, version int64) (
 	return -1, nil
 }
 
+var nextTimeClear = false
```
Comment (enlight) [inline mutable_tree.go]: Doesn't seem to do anything

File: mutable_tree.go [hunk 3]
```diff
@@ -364,11 +366,44 @@ func (tree *MutableTree) GetVersioned(key []byte, version int64) (
...
 // SaveVersion saves a new tree version to disk, based on the current state of
 // the tree. Returns the hash and new version number.
 func (tree *MutableTree) SaveVersion() ([]byte, int64, error) {
+	return tree.saveVersion(true)
+}
+
+func (tree *MutableTree) saveVersion(flushToDisk bool) ([]byte, int64, error) {
 	version := tree.version + 1
 
+	if flushToDisk == true {
```
Comment (enlight) [inline mutable_tree.go]: -> `if flushToDisk {`

File: mutable_tree.go [hunk 4]
```diff
@@ -413,6 +450,21 @@ func (tree *MutableTree) SaveVersion() ([]byte, int64, error) {
 // DeleteVersion deletes a tree version from disk. The version can then no
 // longer be accessed.
 func (tree *MutableTree) DeleteVersion(version int64) error {
+	return tree.DeleteVersionFull(version, true)
+}
+
+func (tree *MutableTree) DeleteVersionFull(version int64, memDeleteAlso bool) error {
+	if tree.memversions[version] == true {
```
Comment (enlight) [inline mutable_tree.go]: -> `if tree.memversions[version]`

File: mutable_tree.go [hunk 5]
```diff
@@ -413,6 +450,21 @@ func (tree *MutableTree) SaveVersion() ([]byte, int64, error) {
...
+		//sometimes you dont want to bother deleting versions in memory
+		if memDeleteAlso == false {
```
Comment (enlight) [inline mutable_tree.go]: -> `if !memDeleteAlso`

File: nodedb.go [hunk 6]
```diff
@@ -72,32 +76,40 @@ func (ndb *nodeDB) GetNode(hash []byte) *Node {
...
 // SaveNode saves a node to disk.
-func (ndb *nodeDB) SaveNode(node *Node) {
+func (ndb *nodeDB) SaveNode(node *Node, flushToDisk bool) {
 	ndb.mtx.Lock()
 	defer ndb.mtx.Unlock()
 
 	if node.hash == nil {
 		panic("Expected to find node.hash, but none found.")
 	}
+	if node.persistedMem == true && flushToDisk == false {
```
Comment (mattkanwisher) [inline nodedb.go]: what happens if we aren't using memversions

Comment (enlight) [inline nodedb.go]: -> `if node.persistedMem && !flushToDisk`

File: nodedb.go [hunk 7]
```diff
@@ -107,11 +119,14 @@ func (ndb *nodeDB) SaveNode(node *Node) {
 	if err := node.writeBytes(buf); err != nil {
 		panic(err)
 	}
-	ndb.batch.Set(ndb.nodeKey(node.hash), buf.Bytes())
-	debug("BATCH SAVE %X %p\n", node.hash, node)
 
-	node.persisted = true
-	ndb.cacheNode(node)
+	if flushToDisk == true {
```
Comment (enlight) [inline nodedb.go]: -> `if flushToDisk {`

File: nodedb.go [hunk 8]
```diff
@@ -132,23 +147,28 @@ func (ndb *nodeDB) Has(hash []byte) bool {
 // NOTE: This function clears leftNode/rigthNode recursively and
 // calls _hash() on the given node.
 // TODO refactor, maybe use hashWithCount() but provide a callback.
-func (ndb *nodeDB) SaveBranch(node *Node) []byte {
+func (ndb *nodeDB) SaveBranch(node *Node, flushToDisk bool) []byte {
 	if node.persisted {
 		return node.hash
 	}
+	if node.persistedMem && flushToDisk == false {
```
Comment (enlight) [inline nodedb.go]: -> `if node.persistedMem && !flushToDisk`

File: nodedb.go [hunk 9]
```diff
@@ -132,23 +147,28 @@ func (ndb *nodeDB) Has(hash []byte) bool {
...
+		node.rightHash = ndb.SaveBranch(node.rightNode, flushToDisk)
 	}
 
 	node._hash()
-	ndb.SaveNode(node)
+	ndb.SaveNode(node, flushToDisk)
 
-	node.leftNode = nil
-	node.rightNode = nil
+	if flushToDisk == true {
```
Comment (enlight) [inline nodedb.go]: -> `if flushToDisk {`

File: nodedb.go [hunk 10]
```diff
@@ -249,7 +274,11 @@ func (ndb *nodeDB) resetLatestVersion(version int64) {
 }
 
 func (ndb *nodeDB) getPreviousVersion(version int64) int64 {
-	itr := ndb.db.ReverseIterator(
+	return ndb.getPreviousVersioni(version, ndb.db)
+}
+
+func (ndb *nodeDB) getPreviousVersioni(version int64, db dbm.DB) int64 {
```
Comment (enlight) [inline nodedb.go]: `getPreviousVersioni` -> `getPreviousVersionFromDB`

**Review by enlight — CHANGES_REQUESTED** [review]

**Review by mattkanwisher — APPROVED** [review]

**Pull request merged**